File: Reversing the equation.java
class Solution
{
   
    String reverseEqn(String S)
    {
        // your code here
        StringBuilder reverseString=new StringBuilder();
        int firstDigit=-1;
        for(int i=S.length()-1;i>=0;i--)
        {
            //System.out.println("Character at Index:: "+i+", is:: "+S.charAt(i));
            if(S.charAt(i)>='0' && S.charAt(i)<='9')
            {
                //System.out.println("In Number Condition");
                if(i>0 && S.charAt(i-1)>='0' && S.charAt(i-1)<='9')
                {
                    //System.out.println("In Next Number Condition");
                    firstDigit=i+1;
                    continue;
                }
                else
                {
                    //System.out.println("In Single Number Condition");
                    if(firstDigit>0)
                    {
                        reverseString.append(S.substring(i,firstDigit));
                        firstDigit=-1;
                    }
                    else
                    {
                        reverseString.append(S.charAt(i));
                    }
                }
            }
            else
            {
                //System.out.println("In Operator Condition");
                if(firstDigit>0)
                {
                    reverseString.append(S.substring(i,firstDigit));
                }
                firstDigit=-1;
                reverseString.append(S.charAt(i));
            }
        }
        return reverseString.toString();
    }
}

File: Reversing the equation.cpp
class Solution
{
  public:
    string reverseEqn (string s)
        {
            //code here.
            string str = "";
            stack<char>st;
            stack<string>sn; int j=0, k=0;
            for(int i=0; i<s.length(); i++)
            {
                if(s[i]=='+' || s[i]=='-' || s[i]=='*' || s[i]=='/')
                {
                    st.push(s[i]);
                    if(!str.empty())
                    sn.push(str);
                    str="";
                    j++;
                    k=0;
                    
                }
                else
                  {  str += s[i];
                    k++;
                    j=0;}
                
            }
            if(!str.empty())
            sn.push(str);
            str = "";
            while(!sn.empty() || !st.empty())
            {
                if(k>j){
                if(!sn.empty())
                {
                    str += sn.top();
                    sn.pop();
                }
                if(!st.empty())
                {
                    str += st.top();
                    st.pop();
                }}
                
                if(k<j){
                if(!st.empty())
                {
                    str += st.top();
                    st.pop();
                }
                if(!sn.empty())
                {
                    str += sn.top();
                    sn.pop();
                }}
                
                
            }
            
            return str;
        }
};
